tools/syz-diff/diff.go

```diff
@@ -0,0 +1,490 @@
+	"github.com/google/syzkaller/pkg/fuzzer/queue"
+	"github.com/google/syzkaller/pkg/instance"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/manager"
+	"github.com/google/syzkaller/pkg/mgrconfig"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/pkg/repro"
+	"github.com/google/syzkaller/pkg/rpcserver"
+	"github.com/google/syzkaller/pkg/signal"
+	"github.com/google/syzkaller/pkg/stat"
+	"github.com/google/syzkaller/pkg/vminfo"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/vm"
+	"github.com/google/syzkaller/vm/dispatcher"
+)
+
+var (
+	flagBaseConfig = flag.String("base", "", "base config")
+	flagNewConfig  = flag.String("new", "", "new config (treated as the main one)")
+	flagDebug      = flag.Bool("debug", false, "dump all VM output to console")
+	flagPatch      = flag.String("patch", "", "a git patch")
+)
+
+func main() {
+	if !prog.GitRevisionKnown() {
+		log.Fatalf("bad syz-diff build: build with make, run bin/syz-diff")
+	}
+	flag.Parse()
+	log.EnableLogCaching(1000, 1<<20)
+
+	baseCfg, err := mgrconfig.LoadFile(*flagBaseConfig)
+	if err != nil {
+		log.Fatalf("base config: %v", err)
+	}
+
+	ctx := vm.ShutdownCtx()
+	base := setup(ctx, "base", baseCfg)
+
+	newCfg, err := mgrconfig.LoadFile(*flagNewConfig)
+	if err != nil {
+		log.Fatalf("new config: %v", err)
+	}
+
+	if *flagPatch != "" {
+		data, err := os.ReadFile(*flagPatch)
+		if err != nil {
+			log.Fatal(err)
+		}
+		extractModifiedFiles(newCfg, data)
+	}
+
+	new := setup(ctx, "new", newCfg)
+	go func() {
+		new.candidates <- manager.LoadSeeds(newCfg, true).Candidates
+	}()
+
+	stream := queue.NewRandomQueue(4096, rand.New(rand.NewSource(time.Now().UnixNano())))
+	base.source = stream
+	new.duplicateInto = stream
+
+	store := &manager.DiffFuzzerStore{BasePath: new.cfg.Workdir}
+	diffCtx := &diffContext{
+		doneRepro:     make(chan *manager.ReproResult),
+		base:          base,
+		new:           new,
+		store:         store,
+		reproAttempts: map[string]int{},
+		http: &manager.HTTPServer{
+			Cfg:       newCfg,
+			StartTime: time.Now(),
+			DiffStore: store,
+		},
+	}
+	diffCtx.http.Pools.Store(new.name, new.pool)
+	diffCtx.http.Pools.Store(base.name, base.pool)
+	new.http = diffCtx.http
+
+	diffCtx.Loop(ctx)
+}
+
+type diffContext struct {
+	store *manager.DiffFuzzerStore
+	http  *manager.HTTPServer
+
+	doneRepro chan *manager.ReproResult
+	base      *kernelContext
+	new       *kernelContext
+
+	mu            sync.Mutex
+	reproAttempts map[string]int
+}
+
+func (dc *diffContext) Loop(ctx context.Context) {
+	reproLoop := manager.NewReproLoop(dc, dc.new.pool.Total()-dc.new.cfg.FuzzingVMs, false)
+	dc.http.ReproLoop.Store(reproLoop)
+	go func() {
+		// Let both base and patched instances somewhat progress in fuzzing before we take
+		// VMs away for bug reproduction.
+		// TODO: determine the exact moment of corpus triage.
+		time.Sleep(15 * time.Minute)
```

Comment:
Do you plan this tool to be used by our customers directly?

Reply:
I plan that it will be usable both as a standalone tool (in its current form) and as a part of the patch fuzzing infra. For the latter we'll need to add some API support to 
a) Determine whether some bug is already known (to determine whether to proceed with it).
b) Report the bugs found in the fuzzed patch.